File: modules/agent/ttsr.go
package agent

import "regexp"

// Rule associates a regex pattern with a rule string to inject when matched.
type Rule struct {
	Pattern *regexp.Regexp
	Rule    string
}

// TTSR (Regex-Triggered Streaming Rule Injection) checks content against a set
// of regex patterns and returns matching rules.
type TTSR struct {
	rules []Rule
}

// NewTTSR creates a new TTSR with the given rules.
func NewTTSR(rules []Rule) *TTSR {
	return &TTSR{rules: rules}
}

// Check tests the content against all rules and returns the rule strings
// for every pattern that matches.
func (t *TTSR) Check(content string) []string {
	var matches []string
	for _, r := range t.rules {
		if r.Pattern.MatchString(content) {
			matches = append(matches, r.Rule)
		}
	}
	return matches
}

File: packages/agent/src/ttsr.ts
/**
 * TTSR (Regex-Triggered Streaming Rule Injection) checks content against a set
 * of regex patterns and returns matching rules.
 * Ported from modules/agent/ttsr.go.
 */

export interface Rule {
  pattern: RegExp;
  rule: string;
}

export class TTSR {
  private _rules: Rule[];

  constructor(rules: Rule[]) {
    this._rules = rules;
  }

  /** Check tests the content against all rules and returns the rule strings
   *  for every pattern that matches. */
  check(content: string): string[] {
    const matches: string[] = [];
    for (const r of this._rules) {
      if (r.pattern.test(content)) {
        matches.push(r.rule);
      }
    }
    return matches;
  }
}
